### PR TITLE
Fade out shortened tab titles + associated performance improvements

### PR DESCRIPTION
**Pull request by Livven**
See commit messages for changes.

(My CSS is very rusty, e.g. I'm not familiar with flexbox but I tried my best to keep everything understandable.)

### CODE REVIEW

**Comment by eoger** [comment]
I'm probably not the right person to review design changes in the 2 first commits.
If a professional UI designer could chime in and give feedback I'd be grateful.

**Comment by eoger** [comment]
I have been able to achieve commit 4 with less code by making `tab-title-wrapper` relative and adding a `position:absolute` div inside of it.
Have a look at the screenshot:
screen shot 2017-06-19 at 8 35 30 pm
What do you think? I may have missed an edge-case...

**Comment by Keith94** [comment]
Great improvements!

Just noticed one of the commits slightly breaks the dark theme.

image

**Comment by Livven** [comment]
@eoger The first two commits simply align the design with the original Tab Center.

The fourth commit also includes a performance optimization. With my setup (several hundred tabs) moving the mouse vertically across the tab list causes the hover effect to noticeably lag behind the mouse location, since showing the tab close button leads to shrinking the `tab-title-wrapper` which requires an expensive layout reflow. With my commit the tab close button is overlaid with `position: absolute` and does not take away space from the `tab-title-wrapper` when shown, thus avoiding the layout reflow and also the laggy behavior.

I probably should've split up that commit by implementing the simpler approach you outlined first and then the performance optimization in another commit.

@Keith94 Forgot about that, will fix.

**Comment by Keith94** [comment]
The sidebar close button uses 4px of border-radius. Should it be made consistent with tab close buttons?

**Comment by Livven** [comment]
@Keith94 Don't think that matters since they close very different things. Just like the window close button is completely square (depending on the platform) while the normal tab close button is circular. Plus this is how it is in the original Tab Center.

**Comment by Livven** [comment]
@eoger @Keith94 Fixed the dark theme.

Also we might want to use CSS variables to separate the rest of the dark theme changes as efficiently as possible, as in 67fdb8a.

File: src/tabcenter.css [hunk 1]
```diff
@@ -363,23 +406,33 @@ img, .tab *:not(.clickable) {
...
+.tab:not(.pinned) > .tab-close {
+    position: absolute;
     display: block;
     width: 24px;
     height: 24px;
-    margin-left: 8px;
+    top: 0;
+    bottom: 0;
+    right: 12px;
+    margin: auto;
```
Comment (eoger) [inline src/tabcenter.css]: How about top: 50% and transform: translateY(-50%).
If my comment is useless feel free to discard it.

**Comment by eoger** [comment]
I'm not a big fan of using absolute positioned elements (I tried to avoid it and used extensively flexbox), but I'm willing to make an exception here if people are having performance problem with the `display:none` switching of close buttons on large tabs lists.

However, one of the extra bonuses your approach brings is animations, have a look:
kapture 2017-06-21 at 0 19 28

If that's not too much to ask, would you mind:
- Merging all your perf improvements/fixes in a single commit in this PR (you can bring the "shortened tab titles" over since it's probably a dependency) and open another PR for the design changes?
This way we can work towards merging the perf improvements asap and ask a designer to review the UI changes.

Thank you for your work, it looks great 👍

File: src/tabcenter.css (hunk 1 above)
Comment (Livven) [inline src/tabcenter.css]: It works the same but I think using values like (0, 0, auto) is cleaner/more expressive/less arbitrary than (50%, -50%). If you disagree feel free to change it!

**Comment by Livven** [comment]
> I'm not a big fan of using absolute positioned elements (I tried to avoid it and used extensively flexbox), but I'm willing to make an exception here if people are having performance problems with the display:none switching of close buttons on large tabs lists.

Agree but it seems like there is no cleaner way of making elements overlap each other which in this case is required for performance. The only alternative I could think of would be artificially increasing the width of the containing element and then using a negative margin on the child element to create the overlap, but that seems more messy if anything.

> However, one of the extra bonuses your approach brings is animations, have a look:

True, will be worth exploring. Unfortunately animating the highlighted background itself doesn't work because the gradient color changes don't animate. To get that would require multiple overlapping gradients and animating the opacity of those.

> Merging all your perf improvements/fixes in a single commit in this PR (you can bring the "shortened tab titles" over since it's probably a dependency)

Done.

> Opening another PR for the design changes.

#74 and #75.

**Comment by Keith94** [comment]
The fading effect sometimes makes the text look 'truncated' at the end without a proper fading look. For example the 'a' in mozilla:

image

Can it be improved slightly to match the horizontal tabs?

**Comment by Livven** [comment]
@Keith94 Thanks for pointing that out, I had fixed that at some point but then undid it again. Fixed again now!

**Comment by Keith94** [comment]
Much better now. 😄
Please also check the pinned tab icon in the latest commit.

**Comment by Livven** [comment]
Thanks and fixed that too! Think that was why I undid the previous fix originally.

**Comment by eoger** [comment]
Awesome work @Livven, it looks and works great, very happy you picked up that work.
I haven't found any other bugs, if our official QA tester @Keith94 😎 can give it a last spin before I merge this?

**Comment by Keith94** [comment]
Haha yes, it works great for me too. I forgot to mention it.

**Comment by eoger** [comment]
Awesome, thanks everyone!

**Pull request merged**